src/renderer/wg_engine/tvgWgGeometry.h

```diff
@@ -24,102 +24,428 @@
+        vdist[0] = 0.0f;
+        vleng[0] = 0.0f;
+        for (size_t i = 1; i < vcount; i++) {
+            vdist[i] = distance(vbuff[i-1], vbuff[i]);
+            vleng[i] = vleng[i-1] + vdist[i];
+        }
+    }
+
+    // close vertex buffer
+    inline void close() {
+        // check if last point is not to close to the first point
+        if (!tvg::zero(distance2(vbuff[0], last())))
+            append(vbuff[0]);
+        closed = true;
+    }
+
+    // append point
+    inline void append(const Point& p) {
+        vbuff[vcount] = p;
+        vcount++;
+    }
+
+    // append source vertex buffer in index range from start to end (end not included)
+    inline void appendRange(const WgVertexBuffer& buff, size_t start_index, size_t end_index) {
+        for (size_t i = start_index; i < end_index; i++)
+            append(buff.vbuff[i]);
+    }
+
+    // append circle (list of triangles)
+    inline void appendCircle(float radius) {
+        // get approx circle length
+        float clen = 2.0f * radius * MATH_PI;
+        size_t nsegs = std::max((uint32_t)(clen / 8), 16U);
+        // append circle
+        Point prev { std::sin(0.0f) * radius, std::cos(0.0f) * radius };
+        for (size_t i = 1; i <= nsegs; i++) {
+            float t = (2.0f * MATH_PI * i) / nsegs;
+            Point curr { std::sin(t) * radius, std::cos(t) * radius };
+            append(Point{0.0f, 0.0f});
+            append(prev);
+            append(curr);
+            prev = curr;
+        }
+    }
 
-    void appendPoint(WgPoint pt);
-    void appendCubic(WgPoint p1, WgPoint p2, WgPoint p3, size_t nsegs = 16);
+    // append cubic spline
+    inline void appendCubic(const Point& v0, const Point& v1, const Point& v2, const Point& v3) {
+        // get approx cubic length
+        float clen = distance(v0, v1) + distance(v1, v2) + distance(v2, v3);
+        size_t nsegs = std::max((uint32_t)(clen / 8), 16U);
+        // append cubic
+        Bezier bezier{v0, v1, v2, v3};
+        for (size_t i = 1; i <= nsegs; i++)
+            append(bezier.at((float)i / nsegs));
+    }
 
-    void trim(WgPolyline* polyline, float trimBegin, float trimEnd) const;
+    // trim source buffer
+    inline void trim(const WgVertexBuffer& buff, float beg, float end) {
+        // empty buffer guard
+        if (buff.vcount == 0) return;
+        // initialize
+        float len_beg = buff.total() * beg;
+        float len_end = buff.total() * end;
+        // find points
+        size_t index_beg = buff.getIndexByLength(len_beg);
+        size_t index_end = buff.getIndexByLength(len_end);
+        float len_total_beg = buff.vleng[index_beg];
+        float len_total_end = buff.vleng[index_end];
+        float len_seg_beg = buff.vdist[index_beg];
+        float len_seg_end = buff.vdist[index_end];
+        // append points
+        float t_beg = 1.0f - (len_total_beg - len_beg) / len_seg_beg;
+        float t_end = 1.0f - (len_total_end - len_end) / len_seg_end;
```

Comment:
to avoid division by 0:
```
        if (index_beg > 0) {
		float t_beg = 1.0f - (len_total_beg - len_beg) / len_seg_beg;
		append(lerp(buff.vbuff[index_beg-1], buff.vbuff[index_beg], t_beg));
	}
        appendRange(buff, index_beg, index_end);
        if (index_end > 0) {
		float t_end = 1.0f - (len_total_end - len_end) / len_seg_end;
		append(lerp(buff.vbuff[index_end-1], buff.vbuff[index_end], t_end));
	}
```